### PR TITLE
Vulnerability Scanner - Prioritize CISA vulnerability content over NVD 

### PR DESCRIPTION
**Pull request by sebasfalcone**
|Related issue|
|---|
| #26098 |

## Description
This PR aims to change the default CNA to **CISA**, because it is a more mature version of the **NVD**

Given that CISA doesn't contain all the CVEs yet, fallback mechanisms are implemented to use the NVD if information is not present in CISA

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp [hunk 1]
```diff
@@ -138,8 +138,8 @@ class DescriptionsHelper final
         }
         else
         {
-            // Fallback to default ADP
-            vendorConfig = TGlobalData::instance().vendorMaps().at(ADP_DESCRIPTIONS_MAP_KEY).at(DEFAULT_ADP);
+            // Fallback to the default CNA
+            vendorConfig = TGlobalData::instance().vendorMaps().at(ADP_DESCRIPTIONS_MAP_KEY).at(DEFAULT_CNA);
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp:142]: Is the comment a little bit confusing or am I misunderstanding what it means? 
The default CNA is CISA, but is not the fallback CNA.

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp [hunk 2]
```diff
@@ -193,13 +193,24 @@ class DescriptionsHelper final
...
-            // Information not from source, try with default ADP
-            databaseFeedManager->getVulnerabiltyDescriptiveInformation(cve, DEFAULT_ADP, descriptionData);
+            // Information not from source, try with default CNA
+            databaseFeedManager->getVulnerabiltyDescriptiveInformation(cve, DEFAULT_CNA, descriptionData);
         }
 
-        if (!descriptionIsReliable() && descriptionSource != DEFAULT_ADP)
+        // Check if the description is reliable
+        // 1. If its not, try with the default CNA
+        // 2. If its still not reliable, try with the fallback CNA
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp:202]: its -> it's

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp [hunk 3]
```diff
@@ -230,9 +241,20 @@ class DescriptionsHelper final
             cvssData.data = descriptionData.data;
         }
 
-        if (!cvssIsReliable() && cvssSource != DEFAULT_ADP)
+        // Check if the CVSS is reliable
+        // 1. If its not, try with the default CNA
+        // 2. If its still not reliable, try with the fallback CNA
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp:246]: its -> it's?

**Comment by MiguelazoDS** [comment]
# Environment 

- Manager: Ubuntu Jammy
- Agents:
  - Windows Server 2019 
  - Arch Linux 
  - Oracle 8

**Comment by MiguelazoDS** [comment]
# Environment 

Due to the changes here: https://github.com/wazuh/intelligence-data/pull/1381/commits/149c9d39bf0ae499d1a54869572fe58643dcd94a I'll replace the oracle agent by a Debian agent. 

So this https://github.com/wazuh/wazuh/pull/26493#issuecomment-2442524318 will be now: 

```yml
- Manager: Ubuntu Jammy
- Agents:
    - Windows Server 2019
    - Arch Linux
    - Debian Bookworm
```
---

# Testing 

- Using offset: 1035590 

image

## Vulnerabilities found 

image

image
image

- Example vulnerability for Arch Linux 

```json
{
  "agent": {
    "id": "001",
    "name": "archlinux",
    "type": "Wazuh",
    "version": "v4.9.0"
  },
  "host": {
    "os": {
      "full": "Arch Linux ",
      "kernel": "6.10.0-arch1-2",
      "name": "Arch Linux",
      "platform": "arch",
      "type": "arch",
      "version": ".rolling"
    }
  },
  "package": {
    "architecture": "x86_64",
    "description": "An easy-to-use, robust and highly configurable VPN (Virtual Private Network)",
    "name": "openvpn",
    "size": 1552249,
    "type": "pacman",
    "version": "2.5.1-1"
  },
  "vulnerability": {
    "category": "Packages",
    "classification": "CVSS",
    "description": "OpenVPN 2.5.1 and earlier versions allows a remote attackers to bypass authentication and access control channel data on servers configured with deferred authentication, which can be used to potentially trigger further information leaks.",
    "detected_at": "2024-10-30T19:59:38.847Z",
    "enumeration": "CVE",
    "id": "CVE-2020-15078",
    "published_at": "2021-04-26T14:15:08Z",
    "reference": "https://community.openvpn.net/openvpn/wiki/SecurityAnnouncements, https://lists.debian.org/debian-lts-announce/2022/05/msg00002.html, https://community.openvpn.net/openvpn/wiki/CVE-2020-15078, https://security.gentoo.org/glsa/202105-25, https://usn.ubuntu.com/usn/usn-4933-1, https://lists.fedoraproject.org/archives/list/package-announce%40lists.fedoraproject.org/message/GJUXEYHUPREEBPX23VPEKMFXUPVO3PMU/, https://lists.fedoraproject.org/archives/list/package-announce%40lists.fedoraproject.org/message/JGEGLC4YGBDN5CGHTNWN2GH6DJJA36T2/, https://lists.fedoraproject.org/archives/list/package-announce%40lists.fedoraproject.org/message/PLDB3OBQ3AODYYRN7NRCABV6I4AUFAT6/",
    "scanner": {
      "source": "Arch Linux Security Tracker",
      "vendor": "Wazuh"
    },
    "score": {
      "base": 5,
      "version": "2.0"
    },
    "severity": "Medium",
    "under_evaluation": false
  },
  "wazuh": {
    "cluster": {
      "name": "jammy"
    },
    "schema": {
      "version": "1.0.0"
    }
  }
}
```

- Example vulnerability for Debian 

```json
{
  "agent": {
    "id": "002",
    "name": "bookworm",
    "type": "Wazuh",
    "version": "v4.9.1"
  },
  "host": {
    "os": {
      "full": "Debian GNU/Linux 12 (bookworm)",
      "kernel": "6.1.0-25-amd64",
      "name": "Debian GNU/Linux",
      "platform": "debian",
      "type": "debian",
      "version": "12"
    }
  },
  "package": {
    "architecture": "amd64",
    "description": "netfilter libip6tc library",
    "name": "libip6tc2",
    "size": 66,
    "type": "deb",
    "version": "1.8.9-2"
  },
  "vulnerability": {
    "category": "Packages",
    "classification": "CVSS",
    "description": "extensions/libxt_tcp.c in iptables through 1.4.21 does not match TCP SYN+FIN packets in --syn rules, which might allow remote attackers to bypass intended firewall restrictions via crafted packets.  NOTE: the CVE-2012-6638 fix makes this issue less relevant.",
    "detected_at": "2024-10-30T20:10:55.248Z",
    "enumeration": "CVE",
    "id": "CVE-2012-2663",
    "published_at": "2014-02-15T14:57:07Z",
    "reference": "https://bugzilla.redhat.com/show_bug.cgi?id=826702, http://www.spinics.net/lists/netfilter-devel/msg21248.html",
    "scanner": {
      "source": "Debian Security Tracker",
      "vendor": "Wazuh"
    },
    "score": {
      "base": 7.5,
      "version": "2.0"
    },
    "severity": "High",
    "under_evaluation": false
  },
  "wazuh": {
    "cluster": {
      "name": "jammy"
    },
    "schema": {
      "version": "1.0.0"
    }
  }
}
```

- Example vulnerability for Windows 

```json
{
  "agent": {
    "id": "003",
    "name": "vagrant",
    "type": "Wazuh",
    "version": "v4.9.1"
  },
  "host": {
    "os": {
      "full": "Microsoft Windows Server 2019 Datacenter Evaluation 10.0.17763.1935",
      "name": "Microsoft Windows Server 2019 Datacenter Evaluation",
      "platform": "windows",
      "type": "windows",
      "version": "10.0.17763.1935"
    }
  },
  "package": {
    "architecture": "i686",
    "name": "Wireshark 2.4.5 32-bit",
    "path": "C:\\Program Files (x86)\\Wireshark",
    "size": 0,
    "type": "win",
    "version": "2.4.5"
  },
  "vulnerability": {
    "category": "Packages",
    "classification": "CVSS",
    "description": "Due to a failure in validating the length provided by an attacker-crafted CP2179 packet, Wireshark versions 2.0.0 through 4.0.7 is susceptible to a divide by zero allowing for a denial of service attack.",
    "detected_at": "2024-10-30T20:20:51.735Z",
    "enumeration": "CVE",
    "id": "CVE-2023-2906",
    "published_at": "2023-08-25T21:15:07Z",
    "reference": "https://gitlab.com/wireshark/wireshark/-/issues/19229, https://takeonme.org/cves/CVE-2023-2906.html, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/6HCUPLDY7HLPO46PHMGIJSUBJFTT237C/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/L4AVRUYSHDNEAJILVSGY5W6MPOMG2YRF/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/TRKHFQPWFU7F3OXTL6IEIQSJG6FVXZTZ/",
    "scanner": {
      "source": "National Vulnerability Database",
      "vendor": "Wazuh"
    },
    "score": {
      "base": 6.5,
      "version": "3.1"
    },
    "severity": "Medium",
    "under_evaluation": false
  },
  "wazuh": {
    "cluster": {
      "name": "jammy"
    },
    "schema": {
      "version": "1.0.0"
    }
  }
}
```

**Comment by sebasfalcone** [comment]
@MiguelazoDS Thanks for testing this out!

## Update
The reason why the execution failed is because the CISA column should be created before the scanning process starts

These columns are created when the databases are generated from the content:
- https://github.com/wazuh/wazuh/blob/36ebc4500b7ddfffec6ee8604f7465a5449c8042/src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/updateCVECandidates.hpp#L247-L250

In order to implement this changes we need at least one CISA CVE for the scanner to generate the corresponding column and prevent this failure

**Review by pereyra-m — CHANGES_REQUESTED** [review]
Hi @sebasfalcone 

GJ!

There are still some things to improve I think:

**Efficacy tests**:  

I've downloaded the artifacts for the QA tests of this PR. 
I find many messages like

```
wazuh-modulesd:vulnerability-scanner:eventDetailsBuilder.hpp:278 handleRequest : Error message: Couldn't find column family: 'descriptions_cisa'

wazuh-modulesd:vulnerability-scanner:eventPackageAlertDetailsBuilder.hpp:213 handleRequest : Error message: [json.exception.out_of_range.403] key 'cisa' not found
```

**Manual tests**:

I've installed this development branch and performed the scans without updating the local map.

I connected a Windows agent, and many errors and warnings are generated.

For example:

```
wazuh-modulesd:vulnerability-scanner[228702] osScanner.hpp:344 at handleRequest(): WARNING: Failed to scan OS: 'windows_10_22h2', CVE Numbering Authorities (CNA): 'nvd', Error: 'Couldn't find column family: 'cisa''.

wazuh-modulesd:vulnerability-scanner[228702] packageScanner.hpp:731 at handleRequest(): WARNING: Failed to scan package: 'Calculator', CVE Numbering Authorities (CNA): 'cisa', Error: 'Couldn't find column family: 'cisa''.
```

The fallback mechanism isn't generating the same results as before this change.

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp (hunk 1 above)
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/descriptionsHelper.hpp:142]: Removed

**Comment by MiguelazoDS** [comment]
# Update 

## Environment 

- Manager: Ubuntu Jammy
- Agents:
    - Windows Server 2019
    - Arch Linux
    - Debian Bookworm
    
## Testing

### Debian Bookworm :red_circle: 

[ossec.tar.gz](https://github.com/user-attachments/files/17680962/ossec.tar.gz)

- Multiple errors messages like this 
```console
2024/11/08 12:43:50 wazuh-modulesd:vulnerability-scanner[39154] eventDetailsBuilder.hpp:277 at handleRequest(): ERROR: Error building event details for CVE: CVE-2010-4756
2024/11/08 12:43:50 wazuh-modulesd:vulnerability-scanner[39154] eventDetailsBuilder.hpp:278 at handleRequest(): ERROR: Error message: Couldn't find column family: 'descriptions_cisa'
```

- It does not index anything 
image
image
image

**Review by GabrielEValenzuela — COMMENTED** [review]
On `decriptionHelper` generates  a segfault when tries to access to callback 
`cvssIsReliable`, and when the data in the FB is NULL, produces the error.

A simple approach could be

```cpp
const auto cvssIsReliable = [&cvssData, &cvssSource]()
        {
           if(cvssData.data->scoreBase()  && cvssData.data->severity())
{
           return false;
}
            if (cvssData.data->scoreBase() < 0.01f || cvssData.data->severity()>str().empty())
            {
                return false;
            }

            return true;
        };
```

**Comment by MiguelazoDS** [comment]
# Debian Bookworm :green_circle: 

## Baseline (4.10.2) 

image

[vulnerabilities.tar.gz](https://github.com/user-attachments/files/17685080/vulnerabilities.tar.gz)

## Development branch (https://github.com/wazuh/wazuh/pull/26493/commits/9f4e0f9f53e8c8293027fe8cef09d77c157bc360)

image
image

[files.tar.gz](https://github.com/user-attachments/files/17685141/files.tar.gz)

No errors or warnings related to the implementation
image

**Comment by sebasfalcone** [comment]
## Failing efficacy tests

### 029

- Before
```json
        {
          "cpes": [
            "cpe:2.3:o:microsoft:windows_server_2008:*:sp2:*:*:*:*:*:*"
          ],
          "defaultStatus": "affected",
          "product": "windows_server_2008",
          "vendor": "microsoft"
        }
```

- After
```json
        {
          "cpes": [
            "cpe:2.3:o:microsoft:windows_server_2008:*:sp2:*:*:*:*:*:*",
            "cpe:2.3:o:microsoft:windows_server_2008:-:*:*:*:*:*:*:*",
            "cpe:2.3:o:microsoft:windows_server_2008:-:gold:*:*:*:*:*:*"
          ],
          "defaultStatus": "unaffected",
          "product": "windows_server_2008",
          "vendor": "microsoft",
          "versions": [
            {
              "lessThanOrEqual": "6.0.6001",
              "status": "affected",
              "version": "0",
              "versionType": "custom"
            },
            {
              "status": "affected",
              "version": "6.0.6001"
            }
          ]
        }
```

**Comment by GabrielEValenzuela** [comment]
# Testing 🔴 

## Steps

- Install indexer and dashboard `4.10.0-1`
- Install manager from branch `4.10.x`
- Check the functionality  of VD




- Update manager with branch enhancement/26098-prioritize-cisa-vulnerability-content-over-nvd-in-the-vulnerability-scanner and commit 9f4e0f9f53e8c8293027fe8cef09d77c157bc360

image


- Init module
- Connect a new agent



- Disable module
- Restart manager
- Enable module
- Restart manager

Here an infinity loop of the indexer connector happens, there's no response from the indexer and `dataQueue` is never empty

**Review by MiguelazoDS — APPROVED** [review]

**Comment by sebasfalcone** [comment]
## Update
I was able to replicate the issue reported by @GabrielEValenzuela installing a centOS 8 agent

I am investigating why this issue appeared in the first place, it doesn't seem related to this PR and its really tied to the OS of the agent being scanned

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osScanner.hpp [hunk 4]
```diff
@@ -87,6 +87,9 @@ class TOsScanner final : public AbstractHandler<std::shared_ptr<TScanContext>>
         {
             try
             {
+                // Update the vulnerability source.
+                data->m_vulnerabilitySource.first = cnaName;
```
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osScanner.hpp:91]: Here we could be assigning the expanded source (ex. redhat 7)

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp [hunk 5]
```diff
@@ -647,6 +636,22 @@ class TPackageScanner final : public AbstractHandler<std::shared_ptr<TScanContex
...
+                          package.name.c_str(),
+                          package.format.c_str(),
+                          package.vendor.c_str(),
+                          cnaName.c_str(),
+                          data->agentName().data(),
+                          data->agentId().data(),
+                          data->agentVersion().data());
+
+                // Update the vulnerability source.
+                data->m_vulnerabilitySource.first = cnaName;
```
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp:653]: Here we could be assigning the expanded source (ex. redhat 7)

**Comment by Dwordcito** [comment]
Rebase this branch

**Pull request closed without merging**